### PR TITLE
[DOCS] Clarify field data cache behavior

### PR DESCRIPTION
**Pull request by wylieconlon**
After talking with @jtibshirani about how the field data cache is used in practice, I have created this PR to clarify all of the small issues I saw in the docs about this:

1. The field data cache docs did not say that global ordinals were included in the cache size
2. The cache docs did not describe the need to clear the cache if the limit is reached
3. The cache docs did not recommend setting an explicit limit https://github.com/elastic/elasticsearch/issues/59829
4. The LRU behavior of the cache under certain conditions was not documented
5. The `_id` field docs previously said they supported aggregation, which was changed in 7.0 https://github.com/elastic/elasticsearch/pull/49166
6. There was a separate page for the `fielddata` mapping parameter, but it was easier to understand by combining with the docs on the `text` field.
7. Lack of cross-linking in general

### CODE REVIEW

File: docs/plugins/mapper-annotated-text.asciidoc [hunk 1]
```diff
@@ -18,7 +18,7 @@ include::install_remove.asciidoc[]
 [[mapper-annotated-text-usage]]
 ==== Using the `annotated-text` field
 
-The `annotated-text` tokenizes text content as per the more common `text` field (see 
+The `annotated-text` tokenizes text content as per the more common <<text, `text` field>> (see
```
Comment (jrodewig) [inline docs/plugins/mapper-annotated-text.asciidoc:21]: Although the source files are in the same repo, this technically is a separate doc book. Across books, we have to use the external link syntax. This should fix the docs build error.
```suggestion
The `annotated-text` tokenizes text content as per the more common {ref}/text.html[`text`] field (see 
```

**Comment by wylieconlon** [comment]
A significant portion of this doc is based on the [limiting memory usage](https://www.elastic.co/guide/en/elasticsearch/guide/2.x/_limiting_memory_usage.html) section of the definitive guide. I think I ended up writing slightly different advice than what @polyfractal wrote there, comparing this text from the guide:

> By default, this setting is unbounded—Elasticsearch will never evict data from fielddata. This default was chosen deliberately... A bounded size forces the data structure to evict data. ... However, with the default settings, the fielddata from the old indices is never evicted! fielddata will just keep on growing until you trip the fielddata circuit breaker (see Circuit Breaker), which will prevent you from loading any more fielddata.

to what I wrote:

> The default cache size is unlimited, causing the cache to grow until it
> reaches the limit set by the field data circuit breaker.
> It is recommended to set a cache size limit that is smaller than the circuit breaker
> value. Setting the limit will cause the cache to behave as a least-recently-updated
> cache, only keeping the most recently requested field data.

It seems like the definitive guide recommends setting a limit on time-based indices, while I was thinking we wanted to recommend it on all indices?

File: docs/reference/mapping/params/fielddata.asciidoc [hunk 2]
```diff
@@ -1,134 +0,0 @@
-[[fielddata]]
-=== `fielddata`
```
Comment (jrodewig) [inline docs/reference/mapping/params/fielddata.asciidoc:2]: It's best practice to add redirects for any removed fields. I added a redirect for this page with 391ab35414f190530d240a554831b17b48ac0803 and c121d75cee297a5cabf41e657e1c740c67287b92

**Comment by polyfractal** [comment]
Haven't looked closely at the PR yet, will be free in a few minutes to take a closer look.  

But a note/context about the Definitive Guide advice: it was written at a time when _most_ aggregations worked via field data.  Doc values were very new at the time and not widely used yet, so most aggregations were purely on-heap in field data and OOMs were common due to people setting sub-optimal field data settings (e.g. a lot of people used the time-based eviction, or tried to set limits which could be ignored in a variety of circumstances due to how things worked, etc).

So just a big caveat to the "old" advice, it's pretty dated at this point and specific to the environment at the time.  So no worries if the advice we write here conflicts with what the DefGuide says. :)

**Review by jtibshirani — COMMENTED** [review]
Thanks @wylieconlon for this contribution, it looks great overall. I mostly left small comments related to wording.

> It seems like the definitive guide recommends setting a limit on time-based indices, while I was thinking we wanted to recommend it on all indices?

We're actually still finalizing the recommendation/ plans for field data cache eviction, so I think it'd be best to avoid giving a strong recommendation in this PR.

**Review by jrodewig — COMMENTED** [review]
Ditto on @jtibshirani's comments. I left some other minor suggestions but nothing major.

Thanks for raising this @wylieconlon. I think moving the field data params under `text` makes a lot of sense and will help users.

File: docs/reference/mapping/fields/id-field.asciidoc [hunk 3]
```diff
@@ -33,12 +33,14 @@ GET my-index-000001/_search
...
 
-The value of the `_id` field is also accessible in aggregations or for sorting,
-but doing so is discouraged as it requires to load a lot of data in memory. In
-case sorting or aggregating on the `_id` field is required, it is advised to
-duplicate the content of the `_id` field in another field that has `doc_values`
-enabled.
-
+The `_id` field is by default not available by default for use with aggregations or sorting.
+To aggregate or sort by the `_id` field, it is recommended to 
+duplicate the `_id` field onto a `keyword` field using the <<copy-to, `copy_to` mapping parameter>>.
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply docs/reference/mapping/fields/id-field.asciidoc:38]: I just realized that it's not possible to use `copy_to` with the `_id` mapper, this only works for custom IDs where the user can manually duplicate the ID into another field.

File: docs/reference/mapping/fields/id-field.asciidoc [hunk 4]
```diff
@@ -3,10 +3,14 @@
...
-<<query-dsl-ids-query,`ids` query>>.
+<<query-dsl-ids-query,`ids` query>>. The `_id` can either be assigned at
+indexing time, or a unique `_id` can be generated by {es}. This field is not
+configurable.
 
-The value of the `_id` field is accessible in certain queries (`term`,
-`terms`, `match`, `query_string`, `simple_query_string`).
+The value of the `_id` field is only accessible in certain queries (`term`,
+`terms`, `match`, `query_string`, `simple_query_string`), but is restricted
+from use in aggregations, sorting, or scripting. For those use cases the
```
Comment (jtibshirani) [inline docs/reference/mapping/fields/id-field.asciidoc:12]: I liked the wording 'duplicate the content', I thought it made the recommendation more clear.

File: docs/reference/mapping/fields/id-field.asciidoc [hunk 5]
```diff
@@ -16,30 +20,23 @@ PUT my-index-000001/_doc/1
   "text": "Document with ID 1"
 }
 
-PUT my-index-000001/_doc/2?refresh=true
+POST my-index-000001/_doc/?refresh=true
```
Comment (jtibshirani) [inline docs/reference/mapping/fields/id-field.asciidoc:23]: I think we should undo this change, it feels weird to be mixing auto-generated and custom IDs ?

File: docs/reference/modules/indices/fielddata.asciidoc [hunk 6]
```diff
@@ -1,28 +1,35 @@
...
+The entries in the cache are expensive to build, so the default behavior is
+to keep the cache loaded in memory. The default cache size is unlimited,
+causing the cache to grow until it reaches the limit set by the <<fielddata-circuit-breaker, field data circuit breaker>>. This behavior can be configured.
+
+If the cache size limit is set, the cache will begin clearing the least-recently-updated
+entries in the cache. This setting can automatically avoid the circuit breaker limit,
+at the cost of rebuilding the cache as needed.
+
+If the circuit breaker limit is reached, further requests that increase the cache
+size will be prevented. In this case you shoul manually <<indices-clearcache, clear the cache>>.
```
Comment (jtibshirani) [inline docs/reference/modules/indices/fielddata.asciidoc:21]: shoul -> should

**Review by jtibshirani — APPROVED** [review]
This looks good to me, just some last minor comments.

File: docs/reference/modules/indices/fielddata.asciidoc [hunk 7]
```diff
@@ -1,28 +1,35 @@
...
 
 `indices.fielddata.cache.size`::
 (<<static-cluster-setting,Static>>)
-The max size of the field data cache, eg `30%` of node heap space, or an
-absolute value, eg `12GB`. Defaults to unbounded.  Also see
-<<fielddata-circuit-breaker>>.
+The max size of the field data cache, eg `38%` of node heap space, or an
+absolute value, eg `12GB`. Defaults to unbounded. Should be set smaller than the
+<<fielddata-circuit-breaker>>, if set.
```
Comment (jrodewig) [inline docs/reference/modules/indices/fielddata.asciidoc:27]: It's not clear whether `if set` applies to the circuit breaker or this setting.
```suggestion
absolute value, eg `12GB`. Defaults to unbounded. This value should not
exceed the <<fielddata-circuit-breaker,`indices.breaker.fielddata.limit`>>.
```

**Comment by polyfractal** [comment]
Whoops, I spaced out and forgot to come back to this PR.  ++ to @jtibshirani's comments, and I think it looks good overall.

I wonder if we should make a more explicit note about how to think about global ordinals?  E.g. if query latency is your top priority, it makes sense to pay the global ord cost up front at indexing time so that users don't hit a query pause while global ords build.  Usually folks who require absolute best query latency also have relatively static indices so this works well.

But if your index is changing rapidly, you end up redoing the same global ordinal work over and over as segments are created/merged.  So this can make it potentially very expensive for indices that are changing a lot (most notably indices with a lot of updates/deletes, and to a lesser-extent append-only).

I'm not sure a good/easy way to phrase that in the docs.  The current text alludes to all the above by describing how it works, but it might be useful to write it out a bit more long form?  Not sure.  I've seen eager-global-ords used as sort of "tweak this, expect miracles" setting when it's really not... just shifting around when/where work is done :)

**Comment by jtibshirani** [comment]
> I wonder if we should make a more explicit note about how to think about global ordinals?

+1 to this suggestion, but maybe we should save it for a follow-up to keep this PR focused on the fielddata cache? The PR scope already has a large scope.

File: docs/reference/mapping/fields/id-field.asciidoc [hunk 8]
```diff
@@ -3,10 +3,14 @@
 
 Each document has an `_id` that uniquely identifies it, which is indexed
 so that documents can be looked up either with the <<docs-get,GET API>> or the
-<<query-dsl-ids-query,`ids` query>>.
+<<query-dsl-ids-query,`ids` query>>. The `_id` can either be assigned at
+indexing time, or a unique `_id` can be generated by {es}. This field is not
+configurable.
```
Comment (jrodewig) [inline docs/reference/mapping/fields/id-field.asciidoc:8]: I don't know if `configurable` is clear here.
```suggestion
<<query-dsl-ids-query,`ids` query>>. You can specify a unique `_id` in the
indexing request or let {es} generate one. However, you can't change an existing
document's `_id`.
```

File: docs/reference/mapping/params/eager-global-ordinals.asciidoc [hunk 9]
```diff
@@ -34,11 +34,10 @@ to be enabled.
 * Operations on parent and child documents from a `join` field, including
 `has_child` queries and `parent` aggregations.
 
-NOTE: The global ordinal mapping is an on-heap data structure. When measuring
-memory usage, Elasticsearch counts the memory from global ordinals as
-'fielddata'. Global ordinals memory is included in the
-<<fielddata-circuit-breaker, fielddata circuit breaker>>, and is returned
-under `fielddata` in the <<cluster-nodes-stats, node stats>> response.
+NOTE: The global ordinal mapping use heap memory as part of the
```
Comment (jrodewig) [inline docs/reference/mapping/params/eager-global-ordinals.asciidoc:37]: `use` -> `uses`
```suggestion
NOTE: The global ordinal mapping uses heap memory as part of the
```

File: docs/reference/mapping/params/eager-global-ordinals.asciidoc [hunk 10]
```diff
@@ -34,11 +34,10 @@ to be enabled.
...
+<<modules-fielddata, field data cache>>. Aggregations on high cardinality fields
+can use a significant amount of heap memory, and could exceed the threshold
+of the <<fielddata-circuit-breaker, field data circuit breaker>>.
```
Comment (jrodewig) [inline docs/reference/mapping/params/eager-global-ordinals.asciidoc:40]: I'd simplify this a little:
```suggestion
can use a lot of memory and trigger the <<fielddata-circuit-breaker, field data
circuit breaker>>.
```

**Review by jrodewig — APPROVED** [review]
LGTM. I left some minor editorial suggestions, but nothing major. Thanks again!

**Comment by polyfractal** [comment]
> +1 to this suggestion, but maybe we should save it for a follow-up to keep this PR focused on the fielddata cache? The PR scope already has a large scope.

Makes sense to me!

File: docs/reference/how-to/search-speed.asciidoc [hunk 11]
```diff
@@ -303,13 +303,14 @@ may become much worse.
...
-Elasticsearch does not know which fields will be used in `terms` aggregations
-and which fields won't. You can tell Elasticsearch to load global ordinals
-eagerly when starting or refreshing a shard by configuring mappings as
-described below:
+<<eager-global-ordinals,Global ordinals>> are a data structure that is used to
+optimize the performance of aggregations. They are calculated lazily and stored in
+the JVM heap as part of the <<modules-fielddata, field data cache>>. For fields
+that are heavily used for bucketing aggregations, you can tell {es} to construct
+and cache the global ordinals before requests are received. This should be done
+carefully because it will increase heap usage and delay indexing until the global
```
Comment (jpountz) [inline docs/reference/how-to/search-speed.asciidoc:311]: It wouldn't delay indexing but refreshing. Indexing can keep going while the index is being refreshed and builds global ordinals.

File: docs/reference/mapping/fields/id-field.asciidoc (hunk 8 above)
Comment (jpountz) [reply docs/reference/mapping/fields/id-field.asciidoc:8]: I think that Wylie means that you con't configure the mapping of the `_id` field as opposed to the value of the `_id` field.

File: docs/reference/modules/indices/fielddata.asciidoc (hunk 7 above)
Comment (jpountz) [inline docs/reference/modules/indices/fielddata.asciidoc:27]: As @jtibshirani mentioned, I'd avoid recommending size-based eviction for this cache.

**Comment by jtibshirani** [comment]
I checked with @wylieconlon, and he was fine with me pushing a last round of fixes then merging.
